org.adempiere.base/src/org/compiere/model/PO.java

```diff
@@ -4982,35 +4982,35 @@ private void checkValidContext() {
 	}
 
 	/*
-	 * To force a cross tenant safe read/write the client program must write code like this:
+	 * To update records belongs to the system tenant from non-system tenant, you must write code like this:
 		try {
-			PO.setCrossTenantSafe();
-			// write here the Query.list or PO.saveEx that is cross tenant safe
+			PO.enterSystemRecordsWritable();
+			// write here code that update records belongs to system tenant
 		} finally {
-			PO.clearCrossTenantSafe();
+			PO.exitSystemRecordsWritable();
 		}
+	   This could be dangerous, use with care.
 	 */
-	private static ThreadLocal<Boolean> isSafeCrossTenant = new ThreadLocal<Boolean>() {
+	private static ThreadLocal<Boolean> isSystemRecordsWritable = new ThreadLocal<Boolean>() {
 		@Override protected Boolean initialValue() {
 			return Boolean.FALSE;
 		};
 	};
-	public static void setCrossTenantSafe() {
-		isSafeCrossTenant.set(Boolean.TRUE);
+	public static void enterSystemRecordsWritable() {
```

Comment:
I still think is good to preserve an option for cross tenant safe readings - I know of implementations that have created a "summarizing" tenant and have extensions that collect data from multiple tenants and summarize (for example for fact_acct table) - we're leaving no option to do that programatically (for swing client for example - is an obsolete piece of code, is easier just to allow the cross-reading than the cost trying to refactor that code)

Reply:
I disagree that we should add an API to make it easy for developer to perform cross tenant operation which is a big security concern in multi tenant, saas environment. For summarizing tenant ( which is not a well define and supported feature in idempiere), you still have the options of view, web service call, direct sql or import/export. Anyway, some customization to code is expected if you are doing something that have not been well defined in the core.

Reply:
I'm not saying to release security checks - we are adding security checks everywhere and we keep adding more to prevent the developers doing wrong things (sometimes prevent completely, sometimes warn, depending on the case).

But still we need to be flexible to allow the developers to do intended things when they know what they are doing, a cross-tenant read is a valid thing in certain scenarios, forbidding totally that option can break existing code out there, and we would be forcing them use JDBC (or fork).

Indeed you're adding an API to write in System, which sounds way more dangerous than cross-tenant reading, and we need to trust that developers use that enterSystemRecordsWritable API knowing what they are doing.

I think our difference of opinions is that you consider any cross-tenant code as wrong - but I think there are cases (uncommon) where that is valid and intended - and I would prefer to be open to such thing.

IMO having setCrossTenantSafe is a good compromise - the developer can do it - but he needs to consciously and explicitly tell the system that he's doing that.
We cannot pretend that iDempiere is protected just dropping the API setCrossTenantSafe, because the developer can simply do that in JDBC and that's something we cannot forbid.

Reply:
The cross-tenant API allows write in System in addition to cross tenant operation and I'm just reducing its scope to only write in System. In fact, it is possible to get rid of it in Core since the only place in core that's using it now is AD_Preference, i.e that can be changed to as an exception for just that table.

Yes, we have the existing JDBC hole in the system (no, it is a huge task but not something impossible to forbid for plugins) but that shouldn't be the reason to add more hole to the system. It would be hard to get rid of a public API once we added it to a release.

Reply:
I have different opinion, JDBC is not a hole.

If JDBC is a hole, then pl/pgsql is a hole too, and the complete SQL language, those are tools to be used when the developer have the proper permissions and I guess somebody must be in charge of giving permissions wisely and review the proper usage of those permissions.

Samely, allowing cross tenant safe readings can be a valid operation, and a required one in certain scenarios.